### ISSUE TITLE
Issue: <Combining LLMChain and ConversationalRelationChain in an agent's routes>

### ISSUE THREAD

**Issue by wolfassi123** [hwchase17/langchain#7997]
### Issue you'd like to raise.

I've been stuck on a task that I was working on, not knowing how to exactly proceed with it. Where I need to combine different type of chain in the routes of an Agent.

So I have managed to build the following pipeline using an Agent. I receive a query and according to the context of the query it gets routed either to any of the different LLM Chains defined (answering Greetings, Goodbyes, Unrelated Questions) or it gets routed to a RetrievalQA that searches for the answer in a VectorStore when the input query is related to the context in question.

Now what I noticed, and have been trying to fix was that sometimes, despite adding memory to my agent, I can receive a follow-up question that will not be routed to the VectorStore as it would labeled as "Unrelated". Example: Supposedly the first question was : "Am I allowed to work remotely on Fridays?". This is related for exampled and the model will reply with no from the VectorStore. Now, if I follow it up with: "How about Mondays?" the model will identify the input query as unrelated as it is not tackling the subject of "Work". Ideally, I would want it to reformulate the question to: "How about working remotely on Monday?" and to route it accordingly to the VectorStore as it is tackling the subject of "work".

I made my research and found that CoversationalRetrievalChains should be part of my solution, but I do not know how to put it exactly in my architecture. The main issue is that LLM Chains expect a certain type of input ('input':x) while ConversationalRelationChain expect another ('question':x, 'chat_history':y).

I tried using LOTR (Lord of all Retrievers) but apparently it hasn't been working as well. I only have 1 retriever (1 VectorStore) and the rest are simple LLMChains. Yet even when I tried LOTR with CoversationalRetrievalChains in order to combine retrievers I am getting the following error: "NotImplementedError: Saving not supported for this chain type."

In short I want to be able to pass the same input, regardless of the route in my Agent, whether it's a LLM Chain or a CoversationalRetrievalChains. I'm a bit lost and would love some help on the subject.

**Comment by shraddhaa26** [comment]
> _Answer generated by a 🤖_
> 
> ## Answer
> I found related issues that might be helpful. I did my best to summarize the solutions, but I recommend looking at the related issues yourself.
> 
> ### Open Issues
> #### [Can't use ConversationalRetrievalChain with router chaining](https://github.com/hwchase17/langchain/issues/7644)
> The suggested solution is:
> 
> The error logs indicate that there are several fields that are either missing or not permitted. The proposed solution suggests using the `MultiRetrievalQAChain` class instead of `MultiPromptChain`. This class supports routing to multiple `BaseRetrievalQA` chains, which is a superclass of `ConversationalRetrievalChain`.
> 
> Here's how you can modify your `generate_router_chain` function to use `MultiRetrievalQAChain`:
> 
> ```python
> def generate_router_chain(prompt_infos, destination_chains, default_chain):
>     """
>     Generates the router chains from the prompt infos.
>     :param prompt_infos The prompt informations generated above.
>     :param destination_chains The LLM chains with different prompt templates
>     :param default_chain A default chain
>     """
>     retriever_infos = [{'name': p['name'], 'description': p['description'], 'retriever': destination_chains[p['name']]} for p in prompt_infos]
>     router_chain = MultiRetrievalQAChain.from_retrievers(llm, retriever_infos, default_chain=default_chain)
>     return router_chain
> ```
> 
> This is just one potential solution and it might need to be adjusted according to your specific needs. If you believe this is a bug that could impact other users, feel free to make a pull request with the necessary changes.
> 
> If this solution doesn't align with your experience, I would recommend upgrading to the latest version of LangChain to ensure you have all the latest features and bug fixes.
> 
> This response is meant to be useful, save you time, and share context. It is **not meant to be a precise solution**, but rather a starting point for your own research.
> 
> Help me be more useful! Please leave a 👍 if this is helpful and 👎 if it is irrelevant.

3 frames
[/usr/local/lib/python3.10/dist-packages/langchain/chains/router/multi_retrieval_qa.py](https://localhost:8080/#) in from_retrievers(cls, llm, retriever_infos, default_retriever, default_prompt, default_chain, **kwargs)
     64             prompt = r_info.get("prompt")
     65             retriever = r_info["retriever"]
---> 66             chain = RetrievalQA.from_llm(llm, prompt=prompt, retriever=retriever)
     67             name = r_info["name"]
     68             destination_chains[name] = chain

[/usr/local/lib/python3.10/dist-packages/langchain/chains/retrieval_qa/base.py](https://localhost:8080/#) in from_llm(cls, llm, prompt, callbacks, **kwargs)
     83         )
     84 
---> 85         return cls(
     86             combine_documents_chain=combine_documents_chain,
     87             callbacks=callbacks,

[/usr/local/lib/python3.10/dist-packages/langchain/load/serializable.py](https://localhost:8080/#) in __init__(self, **kwargs)
     72 
     73     def __init__(self, **kwargs: Any) -> None:
---> 74         super().__init__(**kwargs)
     75         self._lc_kwargs = kwargs
     76 

/usr/local/lib/python3.10/dist-packages/pydantic/main.cpython-310-x86_64-linux-gnu.so in pydantic.main.BaseModel.__init__()

ValidationError: 1 validation error for RetrievalQA
retriever
  Can't instantiate abstract class BaseRetriever with abstract method _get_relevant_documents (type=type_error)


3 frames
[/usr/local/lib/python3.10/dist-packages/langchain/chains/router/multi_retrieval_qa.py](https://localhost:8080/#) in from_retrievers(cls, llm, retriever_infos, default_retriever, default_prompt, default_chain, **kwargs)
     64             prompt = r_info.get("prompt")
     65             retriever = r_info["retriever"]
---> 66             chain = RetrievalQA.from_llm(llm, prompt=prompt, retriever=retriever)
     67             name = r_info["name"]
     68             destination_chains[name] = chain

[/usr/local/lib/python3.10/dist-packages/langchain/chains/retrieval_qa/base.py](https://localhost:8080/#) in from_llm(cls, llm, prompt, callbacks, **kwargs)
     83         )
     84 
---> 85         return cls(
     86             combine_documents_chain=combine_documents_chain,
     87             callbacks=callbacks,

[/usr/local/lib/python3.10/dist-packages/langchain/load/serializable.py](https://localhost:8080/#) in __init__(self, **kwargs)
     72 
     73     def __init__(self, **kwargs: Any) -> None:
---> 74         super().__init__(**kwargs)
     75         self._lc_kwargs = kwargs
     76 

/usr/local/lib/python3.10/dist-packages/pydantic/main.cpython-310-x86_64-linux-gnu.so in pydantic.main.BaseModel.__init__()

ValidationError: 1 validation error for RetrievalQA
retriever
  Can't instantiate abstract class BaseRetriever with abstract method _get_relevant_documents (type=type_error)3 frames
[/usr/local/lib/python3.10/dist-packages/langchain/chains/router/multi_retrieval_qa.py](https://localhost:8080/#) in from_retrievers(cls, llm, retriever_infos, default_retriever, default_prompt, default_chain, **kwargs)
     64             prompt = r_info.get("prompt")
     65             retriever = r_info["retriever"]
---> 66             chain = RetrievalQA.from_llm(llm, prompt=prompt, retriever=retriever)
     67             name = r_info["name"]
     68             destination_chains[name] = chain

[/usr/local/lib/python3.10/dist-packages/langchain/chains/retrieval_qa/base.py](https://localhost:8080/#) in from_llm(cls, llm, prompt, callbacks, **kwargs)
     83         )
     84 
---> 85         return cls(
     86             combine_documents_chain=combine_documents_chain,
     87             callbacks=callbacks,

[/usr/local/lib/python3.10/dist-packages/langchain/load/serializable.py](https://localhost:8080/#) in __init__(self, **kwargs)
     72 
     73     def __init__(self, **kwargs: Any) -> None:
---> 74         super().__init__(**kwargs)
     75         self._lc_kwargs = kwargs
     76 

/usr/local/lib/python3.10/dist-packages/pydantic/main.cpython-310-x86_64-linux-gnu.so in pydantic.main.BaseModel.__init__()

ValidationError: 1 validation error for RetrievalQA
retriever
  Can't instantiate abstract class BaseRetriever with abstract method _get_relevant_documents (type=type_error)

**Comment by shraddhaa26** [comment]
3 frames
[/usr/local/lib/python3.10/dist-packages/langchain/chains/router/multi_retrieval_qa.py](https://localhost:8080/#) in from_retrievers(cls, llm, retriever_infos, default_retriever, default_prompt, default_chain, **kwargs)
     64             prompt = r_info.get("prompt")
     65             retriever = r_info["retriever"]
---> 66             chain = RetrievalQA.from_llm(llm, prompt=prompt, retriever=retriever)
     67             name = r_info["name"]
     68             destination_chains[name] = chain

[/usr/local/lib/python3.10/dist-packages/langchain/chains/retrieval_qa/base.py](https://localhost:8080/#) in from_llm(cls, llm, prompt, callbacks, **kwargs)
     83         )
     84 
---> 85         return cls(
     86             combine_documents_chain=combine_documents_chain,
     87             callbacks=callbacks,

[/usr/local/lib/python3.10/dist-packages/langchain/load/serializable.py](https://localhost:8080/#) in __init__(self, **kwargs)
     72 
     73     def __init__(self, **kwargs: Any) -> None:
---> 74         super().__init__(**kwargs)
     75         self._lc_kwargs = kwargs
     76 

/usr/local/lib/python3.10/dist-packages/pydantic/main.cpython-310-x86_64-linux-gnu.so in pydantic.main.BaseModel.__init__()

ValidationError: 1 validation error for RetrievalQA
retriever
  Can't instantiate abstract class BaseRetriever with abstract method _get_relevant_documents (type=type_error)
